pin-project-internal/src/pin_project/attribute.rs

```diff
@@ -1,621 +1,71 @@
-    token::Comma,
     *,
 };
 
-use crate::utils::*;
+use crate::utils::{SliceExt, CURRENT_PRIVATE_MODULE};
 
 use super::PIN;
 
-pub(super) fn parse_attribute(args: TokenStream, mut item: Item) -> Result<TokenStream> {
-    let cx;
-    let proj_items = match &mut item {
-        Item::Struct(item) => {
-            super::validate_struct(&item.ident, &item.fields)?;
-            cx = Context::new(args, &mut item.attrs, &item.vis, &item.ident, &item.generics)?;
-
-            let packed_check = ensure_not_packed(&item)?;
-            let mut proj_items = cx.parse_struct(&mut item.fields)?;
-            proj_items.extend(packed_check);
-            proj_items
-        }
-        Item::Enum(ItemEnum { attrs, vis, ident, generics, brace_token, variants, .. }) => {
-            super::validate_enum(*brace_token, variants)?;
-            cx = Context::new(args, attrs, vis, ident, generics)?;
-
-            // We don't need to check for '#[repr(packed)]',
-            // since it does not apply to enums.
-            cx.parse_enum(variants)?
-        }
-        _ => {
-            return Err(error!(
-                item,
-                "#[pin_project] attribute may only be used on structs or enums"
-            ));
-        }
-    };
-
-    let mut item = item.into_token_stream();
-    item.extend(proj_items);
-    item.extend(cx.make_unpin_impl());
-    item.extend(cx.make_drop_impl());
-    Ok(item)
+// To generate the correct `Unpin` implementation and the projection methods,
+// we need to collect the types of the pinned fields.
+// However, since proc-macro-attribute is applied before `#[cfg]` and `#[cfg_attr]` on fields,
+// we cannot be collecting field types properly at this timing.
+// So instead of generating the `Unpin` implementation and the projection methods here,
+// delegate their processing to proc-macro-derive.
+//
+// At this stage, only attributes are parsed and the following attributes are
+// added to the attributes of the item.
+// * `#[derive(InternalDerive)]` - An internal helper macro that does the above processing.
+// * `#[pin(#private(#args))]` - Pass the argument of `#[pin_project]` to proc-macro-derive (`InternalDerive`).
+
+pub(super) fn parse_attribute(args: &TokenStream, input: TokenStream) -> Result<TokenStream> {
+    let Input { mut attrs, body } = syn::parse2(input)?;
+
+    let private = Ident::new(CURRENT_PRIVATE_MODULE, Span::call_site());
+    attrs.push(syn::parse_quote! {
+        #[derive(::pin_project::#private::__PinProjectInternalDerive)]
+    });
+    // Use `#private` to prevent users from trying to control `InternalDerive` manually.
+    // `#private` does not guarantee compatibility between patch versions,
+    // so it should be sufficient for this purpose in most cases.
+    attrs.push(syn::parse_quote! {
+        #[pin(#private(#args))]
+    });
+
+    Ok(quote! {
+        #(#attrs)*
+        #body
+    })
 }
 
 #[allow(dead_code)] // https://github.com/rust-lang/rust/issues/56750
```

Comment:
It seems to have been fixed in the latest nightly, but it is left for compatibility with other versions.